### PR TITLE
fix potential data race when report resource group metric

### LINKED ISSUE #8233: MPPTaskScheduleEntry::waitForSchedule null pointer caused tiflash crash
```
[BaseDaemon.cpp:563] [\"\
       0x7715391\\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+124867473]\
                \\tlibs/libdaemon/src/BaseDaemon.cpp:214\
  0x7f514bc6ddf0\\t<unknown symbol> [libc.so.6+347632]\
       0x887d390\\tDB::MPPTaskScheduleEntry::waitForSchedule() [tiflash+143119248]\
                \\tdbms/src/Flash/Mpp/MPPTaskScheduleEntry.cpp:66\
       0x88640a4\\tDB::MPPTask::runImpl() [tiflash+143016100]\
                \\tdbms/src/Flash/Mpp/MPPTask.cpp:538\
       0x1fb48d8\\tauto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()::operator()() [tiflash+33245400]\ 
                \\tdbms/src/Common/wrapInvocable.h:36\
       0x1ea7243\\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+32141891]\
                \\tdbms/src/Common/DynamicThreadPool.cpp:124\
       0x1ea6bd1\\tDB::DynamicThreadPool::fixedWork(unsigned long) [tiflash+32140241]\
                \\tdbms/src/Common/DynamicThreadPool.cpp:138\
       0x1ea88b2\\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(unsigned long), DB::DynamicThreadPool*, unsigned long&>(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&)::'lambda'(auto&&...), DB::DynamicThreadPool*, unsigned long> >(void*) [tiflash+32147634]\
                \\t/usr/local/bin/../include/c++/v1/thread:291\
  0x7f514bcb8802\\tstart_thread [libc.so.6+653314]\"] [source=BaseDaemon] [thread_id=1072]
```

### PR DESCRIPTION
**Pull request by guo-shaoge**
### What problem does this PR solve?

Issue Number: close #8233

Problem Summary: Different MPPTask can report [resource group metric](https://github.com/pingcap/tiflash/blob/master/dbms/src/Flash/Mpp/MPPTaskScheduleEntry.cpp#L66) at the same time, which may cause data race of [adding metric ](https://github.com/pingcap/tiflash/blob/master/dbms/src/Common/TiFlashMetrics.h#L868)

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Common/TiFlashMetrics.h
Comment (JaySon-Huang) [inline dbms/src/Common/TiFlashMetrics.h:1]: I think we should ensure that any component that will report metrics when deconstructing themselves exits before `TiFlashMetrics` gets deconstructed

Comment (guo-shaoge) [reply dbms/src/Common/TiFlashMetrics.h:1]: Is this problem related to this pr?

Comment (JaySon-Huang) [reply dbms/src/Common/TiFlashMetrics.h:1]: OK

File: dbms/src/Common/TiFlashMetrics.h [hunk 1]
```diff
@@ -893,6 +900,7 @@ struct MetricFamily
     prometheus::Family<T> * store_family;
     std::vector<MetricArgType> store_args;
     // <resource_group_name, metrics>
+    std::shared_mutex mu;
     std::unordered_map<String, std::vector<T *>> metrics_map;
```
Comment (SeaRise) [inline dbms/src/Common/TiFlashMetrics.h:904]: Maybe change it to a better name?
Maybe `resource_group_metrics_map` and `resource_group_metrics_mu`

**Review by SeaRise — APPROVED** [review]
Others LGTM

File: dbms/src/Common/TiFlashMetrics.h (hunk 1 above)
Comment (guo-shaoge) [reply dbms/src/Common/TiFlashMetrics.h:904]: done

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**